### PR TITLE
Add option to interpolate volume sampling

### LINKED ISSUE #182: interpolation on/off
## Use Case

linear interpolation of volume samples can lead to visual artifacts that are not desirable especially with segmentations

## Solution

checkbox to switch interpolation on/off.  unfortunately this can not be per channel due to the packing of multi channels into single texture.

### PR DESCRIPTION
**Pull request by toloudis**
Review size estimate: 
medium. - many files are touched but it's not a whole lot of code.  Most of it is just boilerplate for adding this new feature to all the parts of the app.

Description

Resolves #182 

The pathtracer's paths may or may not terminate at the centers of voxels.  By default, we sample the volume using linear filtering so that the sample is a weighted sum of neighboring voxels.

This code change allows an option to switch off the linear sampling and use nearest-neighbor sampling instead.  Basically this makes the rendering look blockier - if you zoom in you can easily see edges of voxels.  This "nearest" lookup can be an improvement for segmentations in which you want to see those actual voxel edges, or even for lower resoluton raw image data.  Unfortunately it has to be a global setting that is applied to all channels because of how the channels are interleaved in GPU memory.

The changes cover:
* adding a checkbox to the GUI
* saving to json, and loading from json, and initializing the GUI when loading
* adding the setting to the Python API, saving to python script, and adding a "Command" that contains the implementation.  The command is similarly added to the javascript API.

The changes are pretty well split up between the three separate commits, if interested.

### CODE REVIEW

File: agave_app/python/RenderInterface.h [hunk 1]
```diff
@@ -70,4 +70,9 @@ class RenderInterface
   virtual int BackgroundColor(float, float, float) = 0;
   virtual int SetIsovalueThreshold(int32_t, float, float) = 0;
   virtual int SetControlPoints(int32_t, std::vector<float>) = 0;
+  virtual int SetBoundingBoxColor(float, float, float) = 0;
+  virtual int ShowBoundingBox(int32_t) = 0;
+  virtual int ShowScaleBar(int32_t) = 0;
+  virtual int SetFlipAxis(int32_t, int32_t, int32_t) = 0;
+  virtual int SetInterpolation(int32_t) = 0;
```
Comment (toloudis) [inline agave_app/python/RenderInterface.h:77]: the above functions were missing, so I just threw them in for consistency. (This class ideally is the complete command api) This code in here is actually dead code currently and probably should be removed in a separate PR.  It was meant as a way to do a cleaner python integration that didn't require you to run agave separately as a server process, but never got it fully working.

File: renderlib/graphics/ImageXyzcGpu.cpp [hunk 2]
```diff
@@ -199,3 +199,12 @@ ImageGpu::updateLutGpu(int channel, ImageXYZC* img)
...
   m_channels[channel].updateLutGpu(channel, img);
 }
+
+void
+ImageGpu::setVolumeTextureFiltering(bool linear)
+{
+  glBindTexture(GL_TEXTURE_3D, m_VolumeGLTexture);
+  glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_MIN_FILTER, linear ? GL_LINEAR : GL_NEAREST);
+  glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_MAG_FILTER, linear ? GL_LINEAR : GL_NEAREST);
+  glBindTexture(GL_TEXTURE_3D, 0);
```
Comment (toloudis) [inline renderlib/graphics/ImageXyzcGpu.cpp:209]: the above is the core implementation of the feature for the pathtracer -  it tells the gpu to sample the volume differently based on the setting

File: renderlib/graphics/gl/Image3D.cpp [hunk 3]
```diff
@@ -253,9 +253,9 @@ Image3D::prepareTexture(Scene& s)
   // glGenTextures(1, &_textureid);
   glBindTexture(GL_TEXTURE_3D, m_textureid);
   check_gl("Bind texture");
-  glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+  glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_MIN_FILTER, useLinearInterpolation ? GL_LINEAR : GL_NEAREST);
   check_gl("Set texture min filter");
-  glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+  glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_MAG_FILTER, useLinearInterpolation ? GL_LINEAR : GL_NEAREST);
```
Comment (toloudis) [inline renderlib/graphics/gl/Image3D.cpp:258]: these changes are the implementation for the "ray march" renderer, same as what the path trace renderer has to do

**Comment by toloudis** [comment]
Taking this back to draft to fix the build... gh-actions build is having trouble with this

File: renderlib/io/CMakeLists.txt [hunk 4]
```diff
@@ -28,22 +28,12 @@ add_subdirectory(libCZI)
...
-FetchContent_MakeAvailable(zstd)
-
 # tensorstore dependency for renderlib
 set(TENSORSTORE_USE_SYSTEM_TIFF ON)
 FetchContent_Declare(
 	tensorstore
-	URL "https://github.com/google/tensorstore/archive/refs/tags/v0.1.58.tar.gz"
-	URL_HASH SHA256=e916c9dc0ca9a7f272fc975b76ea06c452f921e65a8b201fc45b4ea96261a6ad
+	GIT_REPOSITORY https://github.com/google/tensorstore.git
+	GIT_TAG v0.1.61
```
Comment (toloudis) [inline renderlib/io/CMakeLists.txt:36]: The changes in this file are a throw-in to fix builds on github-actions, separate from the intent of the PR.

File: agave_app/python/pyrenderer.h [hunk 5]
```diff
@@ -104,6 +104,8 @@ class OffscreenRenderer : public RenderInterface
   virtual int SetBoundingBoxColor(float, float, float);
   virtual int ShowBoundingBox(int32_t);
   virtual int ShowScaleBar(int32_t);
+  virtual int SetFlipAxis(int32_t, int32_t, int32_t);
```
Comment (meganrm) [inline agave_app/python/pyrenderer.h:107]: is this another adding something that should have been there? I thought the flip axis was an feature you already added

File: agave_pyclient/agave_pyclient/agave.py [hunk 6]
```diff
@@ -937,6 +937,18 @@ def set_flip_axis(self, x: int, y: int, z: int):
...
         self.cb.add_command("SET_FLIP_AXIS", x, y, z)
 
+    def set_interpolation(self, x: int):
+        """
+        Set the volume sampling interpolation mode
+
+        Parameters
+        ----------
+        x: int
+            0 for nearest, 1 for linear
```
Comment (meganrm) [inline agave_pyclient/agave_pyclient/agave.py:947]: just curious: is it just always easier to deal with ints? if I was writing this in typescript I would use the names "nearest" and "linear" so that it was easy to know what the settings were throughout the code. Or have it be a bool, since this feature is just enabling or disabling a method

**Review by meganrm — APPROVED** [review]
seems good, I did just have one question about your parameter choice, but not blocking

File: agave_pyclient/agave_pyclient/agave.py (hunk 6 above)
Comment (toloudis) [reply agave_pyclient/agave_pyclient/agave.py:947]: Yes from an API design standpoint this is not good. 

The reason it's like this is because these commands that come from the python client or the js web client are getting converted to binary data stream to be sent to the c++ code via network (websocket).  And the client/server system I set up can deal with basic types like ints, floats, and arrays of ints and floats.  (Strings are also done as array of bytes)

So when you see these ints-as-booleans creep into the apis, it's just because it keeps things consistent across python, c++, and js codebases with the identical function signature.
A nice-to-have code change would be to set up good symbolic enums for the functions that take data types like this one, or booleans, and convert internally to the integer types. Then the external python and js apis could be more user friendly.

File: agave_app/python/pyrenderer.h (hunk 5 above)
Comment (toloudis) [reply agave_app/python/pyrenderer.h:107]: yes, this was just added for completeness because I noticed it was missing.  Shouldn't even affect any runtime behaviors.

**Review by ascibisz — APPROVED** [review]
I only kinda understand the nearest neighbor vs linear interpolation big picture stuff, but as far as the code doing what you say it should be doing, it looks good to me!

**Comment by toloudis** [comment]
> I only kinda understand the nearest neighbor vs linear interpolation big picture stuff, but as far as the code doing what you say it should be doing, it looks good to me!

For segmentations, which have binary information 0 or 1 everywhere, interpolation can make the edges look fuzzier and less well defined.  It can also have weird visual halo artifacts due to not knowing exactly where the surface is or what direction it's pointing in.
For fuzzier raw data that is more continuous and noisy in nature, the interpolation can look totally fine.
Here's a segmentation image that shows the difference:
With interpolation: (the old way)
testinterp
Without interpolation:
testnointerp
